index.js

```diff
@@ -637,10 +666,47 @@ SendStream.prototype.sendFile = function sendFile(path) {
     fs.stat(p, function (err, stat) {
       if (err) return next(err)
       if (stat.isDirectory()) return next()
-      self.emit('file', p, stat)
-      self.send(p, stat)
+      checkPrecompressionAndSendFile(p, stat)
     })
   }
+
+  function checkPrecompressionAndSendFile(p, stat) {
+    self.emit('file', p, stat)
+    if (!self._precompressionFormats) return self.send(p, stat)
+
+    var state = {
+      contents: [],
+      extensionsToCheck: self._precompressionFormats.length
+    }
+
+    self._precompressionFormats.forEach(function (format) {
+      debug('stat "%s%s"', p, format.extension);
+      fs.stat(p + format.extension, function onstat(err, contentStat) {
```

Comment:
The reason to stat for files that client does not accept: proxies.
Lets assume that our server is behind a caching proxy.
First client sends request without accept-encoding (aka uncompressed). If we do not stat for all the files on the server side even in that case the server will not set the Vary header.
And when a second client comes in with accept-encoding header the proxy that did not see a Vary header will serve the cached uncompressed content.

Thus the downside of precompressed option is that you need to stat for files even for clients that do not request them.